source/common/upstream/wrsq_scheduler.h

```diff
@@ -0,0 +1,170 @@
+#pragma once
+
+#include <algorithm>
+#include <iostream>
+#include <memory>
+#include <queue>
+#include <utility>
+#include <vector>
+
+#include "envoy/common/random_generator.h"
+#include "envoy/upstream/scheduler.h"
+
+#include "common/common/assert.h"
+
+#include "absl/container/flat_hash_map.h"
+
+namespace Envoy {
+namespace Upstream {
+
+// Weighted Random Selection Queue (WRSQ) Scheduler
+// ------------------------------------------------
+// This scheduler keeps a queue for each unique weight among all objects inserted and adds the
+// objects to their respective queue based on weight. When performing a pick operation, a queue is
+// selected and an object is pulled. Each queue gets its own selection probability which is weighted
+// as the sum of all weights of objects contained within. Once a queue is picked, you can simply
+// pull from the top and honor the expected selection probability of each object.
+//
+// Adding an object will cause the scheduler to rebuild internal structures on the first pick that
+// follows. This operation will be linear on the number of unique weights among objects inserted.
+// Outside of this case, object picking is logarithmic with the number of unique weights. Adding
+// objects is always constant time.
+//
+// For the case where all object weights are the same, WRSQ behaves identical to vanilla
+// round-robin. If all object weights are different, it behaves identical to weighted random
+// selection.
+//
+// NOTE: This class only supports integral weights and does not allow for the changing of object
+// weights on the fly.
+template <class C> class WRSQScheduler : public Scheduler<C> {
+public:
+  WRSQScheduler(Random::RandomGenerator& random) : random_(random) {}
+
+  std::shared_ptr<C> peekAgain(std::function<double(const C&)>) override {
+    std::shared_ptr<C> picked{pickAndAddInternal()};
+    if (picked != nullptr) {
+      prepick_queue_.emplace(picked);
+    }
+    return picked;
+  }
+
+  std::shared_ptr<C> pickAndAdd(std::function<double(const C&)>) override {
+    // Burn through the pre-pick queue.
+    while (!prepick_queue_.empty()) {
+      auto prepicked_obj = prepick_queue_.front();
+      prepick_queue_.pop();
+      if (!prepicked_obj.expired()) {
+        return std::shared_ptr<C>(prepicked_obj);
+      }
+    }
+
+    return pickAndAddInternal();
+  }
+
+  void add(double weight, std::shared_ptr<C> entry) override {
+    rebuild_cumulative_weights_ = true;
+    queue_map_[weight].emplace(std::move(entry));
+  }
+
+  bool empty() const override { return queue_map_.empty(); }
+
+private:
+  using ObjQueue = std::queue<std::weak_ptr<C>>;
+  using QueueMap = absl::flat_hash_map<double, ObjQueue>;
+
+  // Used to store a queue's weight info necessary to perform the weighted random selection.
+  struct QueueInfo {
+    double cumulative_weight;
+    double weight;
+    ObjQueue* q;
+  };
```

Comment:
Sorry, I'm not sure I follow. Can you spell it out a bit more?

Reply:
Sure! Since we're using a property of the value as the key, we can either have the same information in two places (what we're currently doing), or use `absl::flat_hash_set` with a custom hash function. That's an improvement on space usage, but then to look up a `QueueInfo` by weight, you'd have to construct a temporary one with the desired weight and check whether that's in the set or something - not great. `absl::flat_hash_set` supports heterogeneous lookup, though, where you can check for presence by providing a key of a different type, as long as it hashes to the same value. See https://abseil.io/tips/144.

Reply:
Sorry, I've wasted a tremendous amount of time trying to guess my way through this and haven't had any luck. Is there some snippet of code you can share that does what you're asking? I think this may be common inside of Google, but externally there's not any example for me to base this off of.

Reply:
Would this help?

https://github.com/envoyproxy/envoy/blob/a0ca08b69974cc0296e242bfb1c2d4e252d832b5/source/common/common/hash.h#L119

Reply:
That does! Thanks, I missed that.

I did want to ask if it's really worth the additional complexity to save the inline double for each unique weight? This scheduler is mostly encountering scenarios where there are not many unique weights per-object, since the performance degrades to be much worse than EDF. We can't use this for things like the least-request LB where there are many many individual weights.

Reply:
Yeah let's punt or drop this. I had forgotten that flat_hash_set imposes constness requirements that we don't want to hack around.